flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -703,27 +707,41 @@ private void archiveGlobalFailure(
             long timestamp,
             CompletableFuture<Map<String, String>> failureLabels,
             Iterable<Execution> executions) {
-        exceptionHistory.add(
+        latestRootExceptionEntry =
                 RootExceptionHistoryEntry.fromGlobalFailure(
-                        failure, timestamp, failureLabels, executions));
+                        failure, timestamp, failureLabels, executions);
+        exceptionHistory.add(latestRootExceptionEntry);
         log.debug("Archive global failure.", failure);
     }
 
     protected final void archiveFromFailureHandlingResult(
             FailureHandlingResultSnapshot failureHandlingResult) {
+        // ALl exceptions as the ConcurrentExceptions when it's not a new attempt.
+        if (!failureHandlingResult.isNewAttempt()) {
+            checkState(latestRootExceptionEntry != null, "It should have old failure.");
+            List<Execution> concurrentlyExecutions = new LinkedList<>();
+            failureHandlingResult.getRootCauseExecution().ifPresent(concurrentlyExecutions::add);
+            concurrentlyExecutions.addAll(failureHandlingResult.getConcurrentlyFailedExecution());
```

Comment:
nit: Here we use a LinkedList on a Set (which is returned by `getConcurrentlyFailedExecution`). Either we switch to `ArrayList` to benefit from the `addAll` call or we revert the return type of `getConcurrentlyFailedExecution` back to `Iterable` and benefit from the single-element add performance of `LinkedList`. WDYT? :thinking: